### ISSUE TITLE
[Data] Best practice for data reuse and locality-aware scheduling in rolling-window LightGBM workloads

### ISSUE THREAD

**Issue by quenlang007-create** [ray-project/ray#65197]
### Description

We would first like to confirm whether Ray already provides a recommended
native solution for this workload. If not, this issue describes the
capabilities that appear to be missing, particularly reuse across highly
overlapping rolling windows, data-local task placement, and memory-aware
admission control for independent LightGBM training tasks.

## Use case

We use Ray to run a large number of independent LightGBM training tasks.

A single Ray Job contains approximately 240 Ray Tasks:

- Each Task loads a continuous 90-day training window.
- Different Tasks use different time windows.
- These are usually rolling windows, so adjacent Tasks have significant overlap in their input data.
- Each Task independently trains one LightGBM model.
- One Job produces approximately 240 models.
- These Tasks do not collaboratively train a single distributed LightGBM model.
- Each LightGBM model is primarily trained with multithreading on one node.

The simplified workload looks like this:

```text
Ray Job
├── Task 1: load 90 days of data → train model 1
├── Task 2: load another 90-day window → train model 2
├── Task 3: load another 90-day window → train model 3
├── ...
└── Task 240: load another 90-day window → train model 240
```

For example, if each window moves forward by one day:

```text
Task 1: Day 1 to Day 90
Task 2: Day 2 to Day 91
Task 3: Day 3 to Day 92
...
```

Adjacent Tasks therefore share 89 days of input data.

## Current cluster and execution model

Our current environment is approximately:

- Ray: 2.56.0
- Python: 3.11.8
- Deployment: KubeRay on Kubernetes
- Cluster size: 3 nodes capable of running training tasks
- Primary training framework: LightGBM

The training dataset is currently small enough to fit into the memory of a single node.

To avoid transferring training data between nodes, the application currently uses the following approach:

1. Load a complete copy of the training dataset onto each of the three nodes.
2. Pin different Jobs to specific nodes.
3. Run all training Tasks of a Job primarily on its assigned node.
4. Avoid cross-node data copies by keeping a complete dataset replica on every node.

This approach is currently stable:

- We have not observed significant Task registration timeout issues.
- We have not encountered node OOM failures.
- Most cross-node data transfers are avoided.
- LightGBM training within a single node is relatively predictable.

However, this approach does not fully use Ray as a distributed data and scheduling system. In practice, it is closer to treating the Ray nodes as independent single-node execution environments.

### 1. The complete dataset may no longer fit into one node

As the training dataset grows, it may eventually exceed the memory capacity of a single node.

At that point, we will no longer be able to keep a complete copy on every node. We may need to:

- Partition the complete dataset into multiple data shards.
- Distribute those shards across multiple nodes.
- Read multiple shards for one 90-day training window.
- Handle windows whose input data spans multiple nodes.

This introduces several questions:

- What is the appropriate data partitioning granularity?
- How can we avoid persistently storing duplicated data for heavily overlapping 90-day windows?
- How can a training Task be scheduled on the node that already contains most of its input data?
- If a window spans multiple nodes, how should cross-node transfer and temporary memory copies be minimized?
- How can we prevent multiple LightGBM Tasks from materializing their inputs simultaneously and causing a node-level OOM?

### 2. Application code currently needs to understand the Ray cluster topology

To bind a Job or Task to a node containing a complete copy of the dataset, our application code currently needs to know:

- Which nodes exist in the Ray cluster.
- Which node contains a particular copy of the dataset.
- Which node a Job should be assigned to.
- The mapping between application Jobs and Ray nodes.

This is manageable with only three nodes.

As the cluster grows, however, this creates several problems:

- Application code becomes coupled to the deployment topology.
- Scaling events and node failures may require changes to node-binding logic.
- Application developers need to understand Ray node identifiers, resources, and scheduling configuration.
- Different Jobs may repeatedly implement similar data-location and Task-scheduling logic.
- Manually maintaining the mapping between Jobs and nodes will become difficult.

Ideally, application code should only need to describe:

```text
This Task requires this date range, these columns, and these training parameters.
```

It should not need to describe:

```text
This Task must run on this specific Ray node.
```

## Desired behavior

Ideally, application code would declare a training window such as:

```python
window = {
    "start_date": "2026-01-01",
    "end_date": "2026-03-31",
    "columns": [...],
}
```

Ray or a data management layer would then automatically:

1. Identify the data shards required by the window.
2. Determine which nodes currently hold those shards.
3. Select a node that has most of the required data and sufficient resources.
4. Schedule the training Task on that node.
5. Transfer only the missing shards.
6. Reuse data shared by overlapping windows.
7. Limit the number of concurrent LightGBM Tasks per node to prevent OOM.
8. Recalculate data placement and Task placement after node scaling or failure.
9. Hide the concrete Ray cluster topology from application code.

### 1. Does Ray already provide a native solution for this workload?

We are aware of several potentially relevant Ray capabilities:

- Ray Core locality-aware scheduling
- Object storage and cross-node transfer through the Ray Object Store
- Ray Data block management, streaming execution, backpressure, and locality scheduling
- Ray Train `LightGBMTrainer`
- Ray Tune concurrency and resource management
- Placement Groups, Node Affinity, Node Labels, and custom resources

However, we are unsure how these capabilities should be combined for this workload.

In particular:

- Can Ray Data automatically manage highly overlapping rolling time windows and reuse data shared by adjacent windows?
- When a training Task depends on multiple `ObjectRef`s or Ray Data blocks, does Ray automatically schedule the Task on the node containing the largest number of local input bytes?
- If the data for one 90-day window is distributed across multiple nodes, what is the recommended Ray execution model?
- For 240 independent LightGBM models, should we use ordinary Ray Tasks, Ray Tune, or Ray Train `LightGBMTrainer`?
- Is `LightGBMTrainer` mainly intended for multiple workers collaboratively training one model, rather than one Job training 240 independent models?
- Does Ray provide a native mechanism to limit per-node training concurrency based on worker heap memory or the actual memory usage of LightGBM?

### 2. Should users manage data partitioning themselves?

If we use Ray Data, would the recommended approach be to:

- Convert the complete dataset into a Ray Dataset.
- Allow Ray Data to generate and manage its blocks.
- Filter the Dataset by date for each training window.
- Convert each window into pandas, NumPy, or `lightgbm.Dataset` inside the training Task.

For a large number of highly overlapping windows, would this repeatedly perform filtering, materialization, and Arrow-to-LightGBM conversion?

Is there a recommended caching or reuse mechanism that avoids independently loading the same underlying data for all 240 windows?

### 3. Should users implement a data-location directory and scheduling layer?

We are exploring a lightweight coordination layer with the following flow:

```text
Canonical time-based data shards
    ↓
Track the nodes holding each shard
    ↓
Combine multiple shards into a logical 90-day window
    ↓
Select the node containing most of the window data
    ↓
Apply per-node LightGBM memory admission control
    ↓
Submit an ordinary Ray Task to train one model
```

This coordination layer might be responsible for:

- Maintaining a data-shard directory.
- Managing cached data and its lifecycle.
- Mapping rolling windows to their underlying shards.
- Selecting the Task execution node based on local data volume.
- Applying per-node LightGBM training memory admission.
- Submitting Tasks in bounded waves instead of starting all 240 Tasks at once.

We would like to understand:

- Would this duplicate functionality already provided by Ray Data or Ray Core?
- Should this be implemented as an extension around Ray Data rather than by independently managing Arrow `ObjectRef`s?
- Is there an existing Ray API for querying the node distribution of `ObjectRef`s or Ray Dataset blocks?
- Is there a recommended way to express both data-location preference and node-memory safety to the Ray scheduler?
- If an application-level coordination layer is necessary, what should the responsibility boundary be between that layer, Ray Core, and Ray Data?

### 4. How should actual LightGBM memory consumption be handled?

Our training input may initially be stored as Arrow data, but LightGBM training usually involves additional materialization:

```text
Arrow
→ pandas / NumPy
→ lightgbm.Dataset
→ LightGBM internal binning and training structures
```

As a result, the peak memory usage of a Task may be significantly larger than the original Arrow input size.

Ray CPU resource declarations can limit the number of concurrent Tasks, but we are unsure about the recommended way to limit LightGBM Task concurrency based on node memory:

```text
Safe concurrency per node
= min(
    concurrency allowed by available CPUs,
    concurrency allowed by available memory
)
```

We would like guidance on the following:

- Does Ray recommend using a custom logical resource to represent training memory?
- Should Ray memory monitoring and OOM prevention be used as the primary concurrency-control mechanism?
- Should bounded concurrency be implemented at the application level using `ray.wait()`?
- Can Ray Data backpressure account for worker heap memory consumed after data is materialized into a `lightgbm.Dataset`?
- Is there a recommended admission-control pattern for workloads with significant and workload-dependent memory amplification?

## Approach we are currently considering

Our current direction is:

1. Partition unique time-series data into reusable Arrow shards.
2. Avoid permanently storing one complete physical copy for every 90-day window.
3. Represent each training window as a logical reference to multiple underlying shards.
4. Prefer nodes containing the largest portion of the required window data.
5. Let Ray Core handle actual Task execution and object transfer.
6. Train one model per Task using single-node multithreaded LightGBM.
7. Limit concurrent training Tasks according to each node’s available memory.
8. Execute the 240 Tasks in bounded waves.
9. Keep node IDs and cluster topology hidden from application code.

At the same time, we are locally exploring a lightweight data scheduling framework. Before investing further, we would like to confirm whether Ray already provides a more native, simpler, or recommended solution.

## Guidance we are looking for

We would appreciate the community’s guidance on the following:

1. What is the recommended Ray stack for this workload?
2. Can Ray Data directly address rolling-window data reuse and data-local scheduling?
3. Should 240 independent LightGBM models be executed using ordinary Ray Tasks, Ray Tune, or Ray Train?
4. Do users need to maintain the locations of data shards themselves?
5. Do users need to implement their own compute-follows-data scheduling?
6. How can data-local scheduling be exposed without making application code aware of cluster topology?
7. How should per-node LightGBM Task concurrency be limited to avoid worker heap OOM?
8. If an extension is required, should it be built around Ray Data, Ray Core scheduling, or an independent data coordination layer?

Our goal is to reuse Ray’s native capabilities wherever possible and avoid rebuilding functionality that Ray already provides.
